subsys/net/ip/tcp2.c

```diff
@@ -832,6 +979,79 @@ static struct tcp *tcp_conn_new(struct net_pkt *pkt)
 	return conn;
 }
 
+static void update_acked_data(struct tcp *conn, struct tcphdr *th)
+{
+	/* If there is any pending data, remove them from re-send as peer
+	 * just acked it.
+	 */
+	u32_t acked_len = th_ack(th) - conn->last_seq;
+	int ret;
+
+	if (!conn->pending_data || acked_len == 0) {
+		return;
+	}
+
+	NET_DBG("Peer acked %u bytes", acked_len);
+
+	net_pkt_cursor_init(conn->pending_data);
+	net_pkt_set_overwrite(conn->pending_data, true);
+
+	/* TODO: Check that pull would succeed beforehand. For this we need
+	 * to know the length of pending_data. But the net_pkt_get_len() is
+	 * expensive, and as we need to know the length information in many
+	 * places, fix this by keeping track of length separately. The length
+	 * calculations in tcp_send_queued_data() would need to be changed too.
+	 */
+	ret = net_pkt_pull(conn->pending_data, acked_len);
+	if (ret < 0 && conn->pending_len != (int)acked_len) {
+		NET_ERR("Trying to ack more data (%d) that was in the "
+			"queue (%u)", acked_len, conn->pending_len);
+		conn->pending_len = 0;
+	} else {
+		conn->pending_len -= acked_len;
+		if (conn->pending_len < 0) {
+			NET_DBG("Pending length underflow by %d",
+				conn->pending_len);
+			conn->pending_len = 0;
+		}
+
+		/* Make sure there is no empty buffers in the beginning
+		 * of the chain as we just pulled data from it (so there
+		 * should not be any)
+		 */
+		if (conn->pending_data->buffer &&
+		    conn->pending_data->buffer->len == 0) {
+			/* TODO: Fix net_pkt_pull() so that this check can
+			 * be removed.
+			 */
+			struct net_buf *buf = conn->pending_data->buffer;
+
+			while (buf) {
+				if (buf->len) {
+					break;
+				}
+
+				conn->pending_data->buffer = buf->frags;
+				buf->frags = NULL;
+
+				net_buf_unref(buf);
+				if (!conn->pending_data->buffer) {
+					break;
+				}
+
+				buf = conn->pending_data->buffer->frags;
+			}
+		}
```

Comment:
This could be perhaps refactored into separate function?

Reply:
sure